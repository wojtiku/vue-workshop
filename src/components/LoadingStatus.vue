<template>
  <div class="container">
    <div v-if="!!isLoading" class="spinner"></div>
    <div v-else-if="!!isError" class="box">
      There was an error while loading the resource, sorry.
    </div>
    <slot v-else />
  </div>
</template>

<script>
  export default {
    props: {
      isLoading: {
        default: false
      },
      isError: {
        default: false
      }
    }
  }
</script>