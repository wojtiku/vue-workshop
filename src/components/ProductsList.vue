<template>
  <div>
    <p v-show="!isLoading && !(isFirstPage && isLastPage)">
      <router-link :to="{ path: '/', query: { page: page - 1 } }" v-show="!isFirstPage" class="btn">Previous page</router-link>
      {{ page }}
      <router-link :to="{ path: '/', query: { page: page + 1 } }" v-show="!isLastPage" class="btn">Next page</router-link>
    </p>

    <LoadingStatus :isLoading="isLoading">
      <ul v-if="products.length" class="product-list">
        <ProductsListItem
          v-for="product in products"
          :key="product.id"
          :product="product"/>
      </ul>
      <p v-else>No products to be shown. Try a different page.</p>
    </LoadingStatus>
  </div>
</template>

<script>
  import {mapActions, mapGetters} from 'vuex';
  import {getAllProducts} from '/src/productService';
  import LoadingStatus from "/src/components/LoadingStatus";
  import ProductsListItem from "/src/components/ProductsListItem";

  export default {
    props: {
      page: Number
    },
    data() {
      return {
        isLoading: true
      }
    },
    created() {
      this.reloadProducts();
    },
    computed: {
      isFirstPage() {
        return this.page === 1;
      },
      isLastPage() {
        return this.products.length === 0;
      },
      ...mapGetters([
        "products"
      ]),
    },
    methods: {
      reloadProducts() {
        this.isLoading = true;
        getAllProducts(this.page)
          .then((data) => this.updateProducts(data))
          .catch((e) => this.updateProducts([]))
          .then(() => this.isLoading = false);
      },
      ...mapActions(["updateProducts"])
    },
    watch: {
      page() {
        this.reloadProducts();
      }
    },
    components: {
      ProductsListItem,
      LoadingStatus
    }
  }
</script>

<style lang="scss" scoped>
  .product-list {
    list-style: none;
    padding: 0;
    margin: 0 0 0 (-$gutter);
    display: flex;
    flex-wrap: wrap;

    &--product {
      $inRow: 4;
      flex-basis: calc(#{100%/$inRow} - #{$gutter});
    }
  }
</style>
