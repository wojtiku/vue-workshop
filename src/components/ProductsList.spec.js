import LoadingStatus from '/src/components/LoadingStatus';
import ProductsList from '/src/components/ProductsList';
import {mockProductList} from '/src/test/mockProduct';
import {createLocalVue, RouterLinkStub, shallowMount} from '@vue/test-utils';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.directive("style-when-broken", ()=>{});

describe('ProductsList.vue', () => {
  let state;
  let actions;
  let getters;
  let store;

  function createWrapper() {
    store = new Vuex.Store({
      state,
      getters,
      actions
    });

    return shallowMount(ProductsList, {
      store,
      localVue,
      stubs: {
        RouterLink: RouterLinkStub,
        LoadingStatus
      }
    });
  }

  beforeEach(() => {
    state = {
      page: 1
    };
    actions = {
      fetchProducts: jest.fn()
    };
    getters = {
      products: () => mockProductList,
      productsStatus: () => ({}),
      currentPageNumber: (state) => state.page
    };
  });

  it('should match snapshot', () => {
    const wrapper = createWrapper();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render product-list-items with correct products', () => {
    const wrapper = createWrapper();
    const items = wrapper.findAll({ name: "ProductsListItem" });

    expect(items).toHaveLength(4);
    expect(items.at(2).props().product).toBe(mockProductList[2]);
  });

  describe("Navigation buttons", () => {
    describe("Back button", () => {
      it('should point to the previous page', () => {
        const wrapper = createWrapper();
        expect(wrapper.findAll(RouterLinkStub).at(0).props().to.query.page).toBe(state.page - 1);
      });

      it('should be hidden on the first page', () => {
        const wrapper = createWrapper();
        expect(wrapper.findAll(RouterLinkStub).at(0).isVisible()).toBe(false);
      });
    });

    describe("Next button", () => {
      it('should point to the next page', () => {
        const wrapper = createWrapper();
        expect(wrapper.findAll(RouterLinkStub).at(1).props().to.query.page).toBe(state.page + 1);
      });

      it('should be hidden when there are no products', () => {
        getters.products = () => [];

        const wrapper = createWrapper();
        expect(wrapper.findAll(RouterLinkStub).at(1).isVisible()).toBe(false);
      });
    });
  });

  describe("Refreshing data", () => {
    it('should fetch products when mounted', () => {
      createWrapper();
      expect(actions.fetchProducts).toHaveBeenCalledTimes(1);
    });

    it('should fetch products when page changes', () => {
      createWrapper();
      state.page += 1;
      expect(actions.fetchProducts).toHaveBeenCalledTimes(2);
    });
  });

  describe("Loading status", () => {
    function getLoadingStatusProps() {
      return createWrapper().find(LoadingStatus).props();
    }

    it('should not show spinner when products are there', () => {
      const props = getLoadingStatusProps();

      expect(props.isLoading).toBeFalsy();
      expect(props.isError).toBeFalsy();
    });

    it('should show spinner when loading', () => {
      getters.productsStatus = () => ({ loading: true });
      const props = getLoadingStatusProps();

      expect(props.isLoading).toBeTruthy();
      expect(props.isError).toBeFalsy();
    });

    it('should show message on error', () => {
      getters.productsStatus = () => ({ error: true });
      const props = getLoadingStatusProps();

      expect(props.isLoading).toBeFalsy();
      expect(props.isError).toBeTruthy();
    });
  })

});