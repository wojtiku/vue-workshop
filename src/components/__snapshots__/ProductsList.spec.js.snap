// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList.vue should match snapshot 1`] = `
<div>
  <p><a class="btn" style="display: none;">Previous page</a>
    1
    <a class="btn">Next page</a></p>
  <div class="container">
    <ul class="product-list">
      <productslistitem-stub product="[object Object]"></productslistitem-stub>
      <productslistitem-stub product="[object Object]"></productslistitem-stub>
      <productslistitem-stub product="[object Object]"></productslistitem-stub>
      <productslistitem-stub product="[object Object]"></productslistitem-stub>
    </ul>
  </div>
</div>
`;
